[HLSL] Implement the f16tof32() intrinsic

tcorringham opened the pull request:
Implement the f16tof32() intrinsic, including DXILand SPIRV codegen, and associated tests.

Fixes #99112

Commit:
[HLSL] Implement the f16tof32() intrinsic

Implement the f16tof32() intrinsic, including DXILand SPIRV codegen,
and associated tests.

s-perron reviewed (commented):
This looks good from an SPIR-V perspective. However, add a comment explain why you don't manually unpack the inputs. It would be fewer spir-v instructions.

tcorringham commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -3780,6 +3751,89 @@ bool SPIRVInstructionSelector::selectResourceNonUniformIndex(
+      GR.getOrCreateSPIRVVectorType(ResEltType, 2, MIRBuilder, false);
+  const TargetRegisterClass *Vec2RegClass = GR.getRegClass(Vec2ResType);
+
+  bool Result = true;
```
I don't have a particular reason to do it this way - it just seemed the most obvious way (it is the first SPIRV code I've touched, so I only have a superficial understanding of the SPIRV codegen).
If you can outline a better way I'll be happy to amend the code - I don't immediately see how 16 bit types help, unless UnpackHalf2x16 isn't used.

tcorringham replied:
As you have probably noticed, I have made changes along the lines of farzonl's comment - moving the expansion of the vector spv_legacyf16tof32 to Clang Codegen and using a spv_unpackhalf2x16 instead.
This simplifies the spirv codegen significantly. I have included a comment along the lines of your suggestion in the clang codegen.

Commit:
Changes from review comments

Move the unrolling of the UnpackHalf2x16 from the SPIRV codegen to
Clang IR CodeGen, and replace the spv_legacyf16tof32 intrinsic with
a spv_unpackhalf2x16 intrinsic.
This greatly simplifies the SPIRV codegen, at the expense of slightly
complicating the Clang IR codegen.

Commit:
Fix clang-format issue

Apply clang-format

Commit:
Update tests for f16tof32()

Replace the f16tof32 test with an unpackhalf2x16 test, to match changes
to the spirv codegen for f16to32 support.
Adjust the f16tof32-errors.hlsl diagnostics test.

tcorringham committed:
Merge branch 'llvm:main' into f16tof32

farzonl reviewed (approved):
LGTM

tcorringham commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -3516,6 +3516,10 @@ bool SPIRVInstructionSelector::selectIntrinsic(Register ResVReg,
     return selectResourceNonUniformIndex(ResVReg, ResType, I);
   }
+  case Intrinsic::spv_unpackhalf2x16: {
+    return selectExtInst(ResVReg, ResType, I, GL::UnpackHalf2x16);
```
I didn't see this until too late - I will make a test and raise a new issue for it.

farzonl replied:
@tcorringham I think you forgot this.
